### PR TITLE
Copy paste

### LINKED ISSUE #100: Copy-Paste

### PR DESCRIPTION
**Pull request by Zeng-Zer**
closes #100

### CODE REVIEW

File: include/InputMethodRelay.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#ifndef INPUTMETHODRELAY_HPP_
```
Comment (Gaspard--) [inline include/InputMethodRelay.hpp]: Please use `#pargma once` instead of a header guard.

File: include/TextInput.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#ifndef TEXTINPUT_HPP_
```
Comment (Gaspard--) [inline include/TextInput.hpp]: see above

File: source/InputMethodRelay.cpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+void InputMethodRelay::handle_input_method(struct wl_listener *listener, void *data)
+{
+  struct wlr_input_method_v2 *input_method = static_cast<struct wlr_input_method_v2*>(data);
+  if (Server::getInstance().seat.getSeat() != input_method->seat)
+    {
+      return;
+    }
+
+  if (input_method != nullptr)
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
  if (input_method)
```

File: include/InputMethodRelay.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  void handle_text_input_enable(struct wl_listener *listener, void *data);
+  void handle_text_input_commit(struct wl_listener *listener, void *data);
+  void handle_text_input_disable(struct wl_listener *listener, void *data);
+  void handle_text_input_destroy(struct wl_listener *listener, void *data);
+
+  void handle_input_method(struct wl_listener *listener, void *data);
+  void handle_input_method_commit(struct wl_listener *listener, void *data);
+  void handle_input_method_destroy(struct wl_listener *listener, void *data);
+
+  TextInput *getTextInput(InputMethodRelay *relay, struct wlr_text_input_v3 *wlr_text_input);
```
Comment (Gaspard--) [inline include/InputMethodRelay.hpp]: Can't find the corresponding function implementation.

File: include/Server.hpp [hunk 5]
```diff
@@ -56,6 +57,7 @@ class Server
   ServerCursor cursor;
   InputManager inputManager;
   Seat seat;
+  InputMethodRelay imr;
```
Comment (Gaspard--) [inline include/Server.hpp]: I'd use a more expressive variable name. `imr` is very vague.

File: include/TextInput.hpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+struct TextInputListeners
+{
+  struct wl_listener pending_focused_surface_destroy;
+};
+
+class TextInput : public TextInputListeners
+{
+public:
+  TextInput(InputMethodRelay& relay, struct wlr_text_input_v3 *wlr_text_input);
```
Comment (Gaspard--) [inline include/TextInput.hpp]:
```suggestion
  TextInput(InputMethodRelay& relay, struct wlr_text_input_v3 *wlr_text_input);
  TextInput(TextInput const &) = delete;
  TextInput(TextInput &&) = delete;
```

File: source/InputMethodRelay.cpp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+					     input->current.text_change_cause);
+  wlr_input_method_v2_send_content_type(input_method,
+					input->current.content_type.hint,
+					input->current.content_type.purpose);
+  wlr_input_method_v2_send_done(input_method);
+}
+
+void InputMethodRelay::handle_text_input_enable(struct wl_listener *listener, void *data)
+{
+  if (input_method == NULL)
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
  if (!input_method)
```

File: source/InputMethodRelay.cpp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  for (auto const& text_input : text_inputs)
+    {
+      if (text_input->input == data)
+	{
+	  if (!text_input->input->current_enabled) {
+	    wlr_log(WLR_INFO, "Inactive text input tried to commit an update");
+	    return;
+	  }
+	  wlr_log(WLR_DEBUG, "Text input committed update");
+	  if (input_method == NULL) {
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
	  if (!input_method) {
```

File: source/InputMethodRelay.cpp [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    }
+}
+
+void InputMethodRelay::handle_text_input_disable(struct wl_listener *listener, void *data)
+{
+  for (auto const& text_input : text_inputs)
+    {
+      if (text_input->input == data)
+	{
+	  if (input_method == NULL)
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
	  if (!input_method)
```

File: source/InputMethodRelay.cpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+void InputMethodRelay::handle_text_input_destroy(struct wl_listener *listener,
+						 void *data) {
+  for (auto text_input = text_inputs.begin(); text_input != text_inputs.end(); ++text_input)
+    {
+      if ((*text_input)->input == data)
+	{
+	  if ((*text_input)->input->current_enabled) {
+	    if (input_method == NULL)
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
	    if (!input_method)
```

File: source/InputMethodRelay.cpp [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	      {
+		wlr_log(WLR_DEBUG, "Disabling text input, but input method is gone");
+		return;
+	      }
+	    wlr_input_method_v2_send_deactivate(input_method);
+	    send_im_done((*text_input)->input);
+	  }
+	  wl_list_remove(&(*text_input)->pending_focused_surface_destroy.link);
+	  wl_list_init(&(*text_input)->pending_focused_surface_destroy.link);
+	  (*text_input)->pending_focused_surface = NULL;
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
	  (*text_input)->pending_focused_surface = nullptr;
```

File: source/InputMethodRelay.cpp [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+	  wlr_text_input_v3_send_done(text_input->input);
+	  break;
+	}
+    }
+}
+
+void InputMethodRelay::handle_input_method_destroy(struct wl_listener *listener, void *data)
+{
+  struct wlr_input_method_v2 *context = static_cast<struct wlr_input_method_v2*>(data);
+  input_method = NULL;
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
  input_method = nullptr;
```

File: source/InputMethodRelay.cpp [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+  for (auto const& text_input : text_inputs)
+    {
+      if (text_input->pending_focused_surface)
+	{
+	  wlr_text_input_v3_send_enter(text_input->input,
+				       text_input->pending_focused_surface);
+	  wl_list_remove(&text_input->pending_focused_surface_destroy.link);
+	  wl_list_init(&text_input->pending_focused_surface_destroy.link);
+	  text_input->pending_focused_surface = NULL;
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
	  text_input->pending_focused_surface = nullptr;
```

**Review by Gaspard-- — CHANGES_REQUESTED** [review]
Quick first "obvious fixes" pass.

File: include/TextInput.hpp (hunk 6 above)
Comment (Zeng-Zer) [reply include/TextInput.hpp]: This is actually not possible because i store unique_ptr of TextInput in a vector

Comment (Gaspard--) [reply include/TextInput.hpp]: Should be possible if you didn't move the data into unique_ptr, going to check.

File: source/InputMethodRelay.cpp [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+	}
+    }
+}
+
+void InputMethodRelay::handle_text_input([[maybe_unused]]struct wl_listener *listener,
+					 [[maybe_unused]]void *data)
+{
+  struct wlr_text_input_v3 *input = static_cast<struct wlr_text_input_v3*>(data);
+
+  text_inputs.emplace_back(std::make_unique<TextInput>(TextInput(*this, input)));
```
Comment (Gaspard--) [inline source/InputMethodRelay.cpp]:
```suggestion
  text_inputs.emplace_back(std::make_unique<TextInput>(*this, input));
```
Does this work?

**Review by Gaspard-- — APPROVED** [review]
LGTM

**Comment by Gaspard--** [comment]
Rebased into master, see 44dd1c8af75792eccd5f13f9a7eda6cd05dafe99

**Pull request closed without merging**